diff --git a/lib/client-parser.c b/lib/client-parser.c
index b979c62..79b25be 100644
--- a/lib/client-parser.c
+++ b/lib/client-parser.c
@@ -439,6 +439,7 @@ spill:
 
 			/* get it sent as soon as possible */
 			lws_callback_on_writable(wsi);
+			callback_action = LWS_CALLBACK_CLIENT_RECEIVE_PING;
 ping_drop:
 			wsi->u.ws.rx_ubuf_head = 0;
 			handled = 1;
@@ -494,7 +495,7 @@ ping_drop:
 		 * It's nicely buffered with the pre-padding taken care of
 		 * so it can be sent straight out again using lws_write
 		 */
-		if (handled)
+		if (handled && callback_action != LWS_CALLBACK_CLIENT_RECEIVE_PING)
 			goto already_done;
 
 		eff_buf.token = &wsi->u.ws.rx_ubuf[LWS_PRE];
@@ -533,7 +534,7 @@ utf8_fail:			lwsl_info("utf8 error\n");
 		}
 
 		if (eff_buf.token_len < 0 &&
-		    callback_action != LWS_CALLBACK_CLIENT_RECEIVE_PONG)
+		    callback_action != LWS_CALLBACK_CLIENT_RECEIVE_PONG && callback_action != LWS_CALLBACK_CLIENT_RECEIVE_PING)
 			goto already_done;
 
 		if (!eff_buf.token)
diff --git a/lib/libwebsockets.h b/lib/libwebsockets.h
index 23d6588..2081b0a 100644
--- a/lib/libwebsockets.h
+++ b/lib/libwebsockets.h
@@ -1235,6 +1235,8 @@ enum lws_callback_reasons {
 	/**< RAW mode file is writeable */
 	LWS_CALLBACK_RAW_CLOSE_FILE				= 66,
 	/**< RAW mode wsi that adopted a file is closing */
+	LWS_CALLBACK_CLIENT_RECEIVE_PING			=  67,
+	/**< clients receive PING packets with this callback reason */
 
 	/****** add new things just above ---^ ******/
 
diff --git a/lib/ssl.c b/lib/ssl.c
index 651757f..d928afc 100644
--- a/lib/ssl.c
+++ b/lib/ssl.c
@@ -378,6 +378,9 @@ lws_ssl_capable_read(struct lws *wsi, unsigned char *buf, int len)
 	lwsl_debug("%p: SSL_read says %d\n", wsi, n);
 	/* manpage: returning 0 means connection shut down */
 	if (!n) {
+		/* Setting close reason for SSL socket close from server */
+		wsi->vhost->protocols->callback(wsi, LWS_CALLBACK_WS_PEER_INITIATED_CLOSE,
+                                              wsi->user_space, "SSL_Socket_Close", 16 );
 		n = lws_ssl_get_error(wsi, n);
 		lwsl_debug("%p: ssl err %d errno %d\n", wsi, n, errno);
 		if (n == SSL_ERROR_ZERO_RETURN)
